File: apps/desktop/src-tauri/src/models/task.rs
use serde::{Deserialize, Serialize};

use super::tag::Tag;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Task {
    pub id: String,
    pub list_id: String,
    pub parent_task_id: Option<String>,
    pub title: String,
    pub content: Option<String>,
    pub priority: i32,
    pub status: i32,
    pub start_date: Option<String>,
    pub due_date: Option<String>,
    pub due_timezone: Option<String>,
    pub recurrence_rule: Option<String>,
    pub sort_order: i32,
    pub heading_id: Option<String>,
    pub completed_at: Option<String>,
    pub created_at: String,
    pub updated_at: String,
    pub deleted_at: Option<String>,
    pub scheduled_start: Option<String>,
    pub scheduled_end: Option<String>,
    pub estimated_minutes: Option<i32>,
    pub subtasks: Vec<Task>,
    pub tags: Vec<Tag>,
}

File: services/server/internal/models/task.go
package models

import (
	"time"

	"github.com/google/uuid"
)

// Task represents a task in the Cross 2 application.
type Task struct {
	ID             uuid.UUID  `json:"id"`
	UserID         uuid.UUID  `json:"userId"`
	ListID         uuid.UUID  `json:"listId"`
	ParentTaskID   *uuid.UUID `json:"parentTaskId"`
	Title          string     `json:"title" validate:"required,max=500"`
	Content        *string    `json:"content"`
	Priority       int        `json:"priority" validate:"gte=0,lte=3"`
	Status         int        `json:"status"`
	DueDate        *time.Time `json:"dueDate"`
	DueTimezone    *string    `json:"dueTimezone"`
	RecurrenceRule *string    `json:"recurrenceRule"`
	SortOrder      int        `json:"sortOrder"`
	CompletedAt    *time.Time `json:"completedAt"`
	CreatedAt      time.Time  `json:"createdAt"`
	UpdatedAt      time.Time  `json:"updatedAt"`
	DeletedAt      *time.Time `json:"deletedAt"`
	Subtasks       []Task     `json:"subtasks"`
	Tags           []Tag      `json:"tags"`
}

// TaskUpdatePayload contains optional fields for partially updating a task.
type TaskUpdatePayload struct {
	ListID         *uuid.UUID `json:"listId"`
	ParentTaskID   *uuid.UUID `json:"parentTaskId"`
	Title          *string    `json:"title"`
	Content        *string    `json:"content"`
	Priority       *int       `json:"priority"`
	Status         *int       `json:"status"`
	DueDate        *time.Time `json:"dueDate"`
	DueTimezone    *string    `json:"dueTimezone"`
	RecurrenceRule *string    `json:"recurrenceRule"`
	SortOrder      *int       `json:"sortOrder"`
}
